authorization/policy.go

```diff
@@ -34,23 +36,16 @@ func (p ActionResourcePolicy) Authorize(ctx context.Context, userID string, auth
 		}
 	}
 
-	if ok, err := authorizer.IsAuthorizedWithContext(ctx, userID, p.Action, resource); err != nil {
-		switch status.Code(err) {
-		case codes.Canceled:
-			return context.Canceled
-		}
-
-		return fmt.Errorf("unable to call IsAuthorizedWithContext: %w", err)
-	} else if !ok {
-		if _, err := authorizer.GetResourceWithContext(ctx, resource.Id, resource.Type); err != nil {
-			switch status.Code(err) {
-			case codes.Canceled:
-				return context.Canceled
-			case codes.NotFound:
-				return custom_problems.ResourceNotFound(resource.Id, resource.Type)
-			}
+	ok, reason, err := authorizer.IsAuthorizedWithReasonWithContext(ctx, userID, p.Action, resource)
+	if status.Code(err) == codes.Canceled {
+		return context.Canceled
+	} else if err != nil {
+		return fmt.Errorf("unable to call IsAuthorizedWithReasonWithContext: %w", err)
+	}
 
-			return fmt.Errorf("unable to call GetResourceWithContext: %w", err)
+	if !ok {
+		if reason == authorize.ReasonResourceNotFound {
+			return custom_problems.ResourceNotFound(resource.Id, resource.Type)
 		}
 
 		return custom_problems.Unauthorized(userID, custom_problems.PolicyViolation{
```

Comment:
We do have reason ReasonAccessDenied. We can return direct reason.

Reply:
Since you implemented the reason as a string we can not return the reason string directly.

The wanted behavior is to always return Unauthorized if the authorized deny the request, except when the reason was that the requested resource did not exist. Do you have plans to add more reasons?

Reply:
As of now there is no plan to implement more reason. But we can change that in authorise to appropriately return the reason. that way we can have a clean code in middleware.

Reply:
I do not understand what you want to change.

Reply:
no worries.

Reply:
@kulbhushan-skf : the Authorize service shouldn't return `Problems`, it will return string so that the caller can figure out how to translate it. Otherwise we would need to create a map of reasons to HTTP status codes as well to know which types to return. If more reason codes are added, they'll by default return Unauthorized until we customize them here (which is a good starting behavior). This should stay simple as it is for now, then we can expand later if required.